library/system/io.lean

```diff
@@ -68,6 +91,14 @@ interface.iterate e α a f
 def forever {e} (a : io_core e unit) : io_core e unit :=
 iterate () $ λ _, a >> return (some ())
 
+def catch {e₁ e₂ α} (a : io_core e₁ α) (b : e₁ → io_core e₂ α) : io_core e₂ α :=
+interface.catch e₁ e₂ α a b
+
+instance : alternative io :=
+{ interface.monad _ with
+  orelse := λ _ a b, catch a (λ _, b),
+  failure := λ _, io.fail "failure" }
+
```

Comment:
> Note that the alternative laws are upheld for non-backtracking parsers,

AFAICT they are only upheld by the *backtracking* combinators (as in attoparsec and your parser combinators, and as opposed to parsec).

In the non-backtracking combinators you need to keep track of whether input has been consumed (that is, if we should backtrack or not).  Hence you necessarily have multiple kinds of failure that you can distinguish.  In particular `a >> fail` can be different from `fail`, since you can observe whether `a` consumes input or not.

Generalizing this point, I think that all parser combinators *should violate* the `a >> fail = a` property.  This is because we want good error messages:
```lean
run (str "foo" >> fail) "foobar"
-- error at foo[]bar
run fail "foobar"
-- error at []foobar
```
Hence `str "foo" >> fail ≠ fail`.